Bug 1685123 docs for implementation of WebEx manifest sandbox (#44709)
mdn/content: files/en-us/mozilla/add-ons/webextensions/manifest.json/sandbox/index.md

---
title: sandbox
slug: Mozilla/Add-ons/WebExtensions/manifest.json/sandbox
page-type: webextension-manifest-key
browser-compat: webextensions.manifest.sandbox
sidebar: addonsidebar
---

<table class="fullwidth-table standard-table">
  <tbody>
    <tr>
      <th scope="row">Type</th>
      <td><code>Object</code></td>
    </tr>
    <tr>
      <th scope="row">Mandatory</th>
      <td>No</td>
    </tr>
    <tr>
      <th scope="row">Manifest version</th>
      <td>2 or higher</td>
    </tr>
    <tr>
      <th scope="row">Example</th>
      <td>
        <pre class="brush: json">
"sandbox": {
 "pages": ["sandbox.html"]
}</pre>
      </td>
    </tr>
  </tbody>
</table>

Use the `sandbox` key to designate one or more of an extension's pages as **sandboxed pages**.

> Sandboxed pages are loaded with a unique, opaque origin, instead of the extension's usual `moz-extension://` origin. As a result:

Reviewer: Link "opaque origin" to https://developer.mozilla.org/en-US/docs/Glossary/Origin#opaque_origin

Also add a bullet point mentioning that web platform APIs bound to the origin are unavailable. For some examples, see https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Content-Security-Policy/sandbox#allow-same-origin

Author: Done

- Sandboxed pages can't access [WebExtension APIs](/en-US/docs/Mozilla/Add-ons/WebExtensions/API). The `browser` and `chrome` global objects are not available.

> - Sandboxed pages can't access, and can't be accessed by, other pages in the extension, except by using {{DOMxRef("Window.postMessage()")}}.

Reviewer: This statement is too firm. There are other ways to communicate. Suggested edit:

```suggestion
- Sandboxed pages can't access, and can't be accessed by, other pages in the extension, except indirectly by communicating through APIs such as {{DOMxRef("Window.postMessage()")}}.
```

A sandboxed page can be given a more permissive [content security policy (CSP)](#content_security_policy_for_sandboxed_pages) than the rest of the extension. This includes a CSP that permits [`eval()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/eval) and similar constructs that are blocked by an extension's [default content security policy](/en-US/docs/Mozilla/Add-ons/WebExtensions/Content_Security_Policy#default_content_security_policy). Because a sandboxed page can't use WebExtension APIs or reach the rest of the extension directly, this can be done without weakening the security of the extension as a whole.

> This makes the `sandbox` key useful for including a third-party library that relies on `eval()` or `new Function()`, such as some templating engines: load the library in a sandboxed page, and use `postMessage()` to send it data from, and return results to, the rest of the extension.

Reviewer: The framing of template engine is likely copied from Chrome. Although that was a motivating factor back then (10+ years ago), it is hardly a relevant considerations now. These days such libraries are mostly designed to work even with a strict CSP. Let's drop the mention of templating libraries (the general reference to eval/Function are sufficient).

Author: Done

> ## Manifest V2 syntax

Reviewer: Wouldn't it be less confusing if we:
- declare one overview, and emphasize for the sandbox.content_security_policy key that it is MV2 only
- at that description, point to the section on content_security_policy.sandbox (on the same page) for the MV3 syntax
- note that the `content_security_policy.sandbox` is MV3-only, and that for MV2 the other key has to be used?

This would enable us to add new manifest keys to sandbox without being forced to duplicate that documentation in two sections. I anticipate that we will likely have a new manifest sub key in `sandbox` as part of https://bugzilla.mozilla.org/show_bug.cgi?id=2053336

Author: Good point, done

      </td>
    </tr>
    <tr>
      <td><code>content_security_policy</code></td>
      <td><code>String</code></td>
      <td>
 Optional. The content security policy applied to the sandboxed pages. See <a href="#content_security_policy_for_sandboxed_pages">Content security policy for sandboxed pages</a>.
      </td>
    </tr>
  </tbody>
</table>

```json
"sandbox": {
  "pages": ["sandbox.html"],
  "content_security_policy": "sandbox allow-scripts; script-src 'self' 'unsafe-eval';"
}
```

## Manifest V3 syntax

In Manifest V3, `sandbox` only supports the `pages` property:

```json
"sandbox": {
  "pages": ["sandbox.html"]
}
```

The content security policy for sandboxed pages is instead set using the `sandbox` property of the [`content_security_policy`](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/content_security_policy) key:

```json
"content_security_policy": {
  "sandbox": "sandbox allow-scripts; script-src 'self' 'unsafe-eval';"
}
```

## Content security policy for sandboxed pages

If a policy is not supplied, sandboxed pages get this default content security policy:

```plain
sandbox allow-scripts; script-src 'self';
```

> This isolates a sandboxed page from the rest of the extension, but doesn't allow `eval()` or similar constructs. To permit these, include `'unsafe-eval'` (or `'wasm-unsafe-eval'` for [WebAssembly](/en-US/docs/WebAssembly)) in the `script-src` directive of a custom policy.

Reviewer: ```suggestion
This isolates a sandboxed page from the rest of the extension, but doesn't allow `eval()` or similar constructs. To permit these, include `'unsafe-eval'` in the `script-src` directive of a custom policy.
```

Dropped wasm mention to avoid giving the wrong impression that wasm requires sandboxed pages.

> [!NOTE]

> > Chrome has a more permissive default sandboxed pages CSP: `sandbox allow-scripts allow-forms allow-popups allow-modals; script-src 'self' 'unsafe-inline' 'unsafe-eval'; child-src 'self';`.

Reviewer: This is incorrect; Chrome's docs currently state this but it is inaccurate.

Chrome's current default sandbox CSP is `sandbox allow-scripts; script-src 'self'`
following discussion at https://github.com/w3c/webextensions/issues/98
and explicitly stated at https://docs.google.com/document/d/1m82BcCPWDTEzy1R6gbCbUwe-ma-bCTTF_JwRnt3O_IA/edit?tab=t.0#heading=h.oj1dbds5d377

```suggestion
```

Any custom policy supplied for sandboxed pages must meet these requirements:

- It must include the {{CSP("sandbox")}} directive.
- The {{CSP("sandbox")}} directive must not include the `allow-same-origin` keyword. Allowing this would give the page access to the rest of the extension's origin, defeating the purpose of sandboxing it.

## Example

This example loads a third-party templating library that uses `eval()`-like constructs into a sandboxed page, and uses it from a popup.

`manifest.json`:

```json
{
  "manifest_version": 3,
  "name": "Sandbox example",
  "version": "1.0",
  "action": {
    "default_popup": "popup.html"
  },
  "sandbox": {
    "pages": ["sandbox.html"]
  },
  "content_security_policy": {
    "sandbox": "sandbox allow-scripts; script-src 'self' 'unsafe-eval';"
  }
}
```

`sandbox.html`:

```html
<!doctype html>
<html lang="en">
  <head>
    <meta charset="utf-8" />

>     <script src="templating-library.js"></script>

Reviewer: Is this intended as a placeholder example? The example on the page itself does not run because the definition of the file is missing.

Author: It's now a fully working example. However, after debugging an issue I found with Claude, it responded that "sandbox.js rejected every message with if (event.origin !== location.origin) — but a sandboxed page has an opaque origin (serializes to "null"), while the sender (popup.html) has a real moz-extension:// origin, so that check could never pass and the render would never happen. I replaced it with event.source !== window.parent (and added the matching check on the popup.js side, event.source !== sandbox.contentWindow), which validates the sender by window identity instead of origin — this is what MDN's own postMessage() security-concerns guidance recommends specifically because origin comparisons don't work for opaque-origin frames."

Reviewer: Checking `event.source` only is insufficient. If a page is listed in web_accessible_resources, any web page can embed the page and try to community with it.

Your finding with `event.origin` only applies to messages from the sandboxed page. Indeed, in that case the origin is "null".
BUT messages coming from the extension page still have origin `moz-extension://uuid` (or `chrome-extension://id` - this behavior is standard and cross-browser), which can be compared with `location.origin` (since the sandboxed page's URL is still an extension URL, as stated in the latest update on fingerprinting).

    <script src="sandbox.js"></script>
  </head>
  <body></body>
</html>
```

`sandbox.js` listens for messages from the popup, renders a template using the sandboxed library, and posts the result back:

```js

> window.addEventListener("message", (event) => {

Reviewer: Below I am showing a concrete snippet to validate messages. Can you include that here too, and prominently mention a link to the security-relevant best practices at https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage#security_concerns

Ideally we should have a dedicated section on this manifest page that clarifies how this feature should be used:
- messages between the extension and sandboxed page should be strictly validated, with link to https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage#security_concerns
- don't execute untrusted code. Although the extension does not have access to privileged APIs, the extension URL may identify the use of the extension and be abused for fingerprinting purposes.
- don't use sandboxed pages with a relaxed CSP if your only objective is the use of WebAssembly. Instead, add `'wasm-unsafe-eval'` to the extension's `content_security_policy` instead.

Author: Done

  const { template, context } = event.data;
  const render = TemplatingLibrary.compile(template);
  event.source.postMessage({ result: render(context) }, event.origin);
});
```

`popup.html` embeds the sandboxed page in a hidden iframe:

```html
<!doctype html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <script src="popup.js"></script>
  </head>
  <body>
    <iframe id="sandbox" src="sandbox.html" hidden></iframe>
  </body>
</html>
```

`popup.js` sends data to the sandboxed page and handles the result:

```js
const sandbox = document.getElementById("sandbox");

sandbox.addEventListener("load", () => {
  sandbox.contentWindow.postMessage(
    { template: "Hello, {{name}}!", context: { name: "world" } },
    "*",
  );
});

> window.addEventListener("message", (event) => {

Reviewer: As security best practice, always validate the source of the message:

```suggestion
window.addEventListener("message", (event) => {
  if (event.origin !== location.origin) {
    // Reject messages not coming from the extension.
    // Although the popup origin is opaque,
    // location.origin still reflects the true URL of
    // the resource in the extension package.
    return;
  }
```

This makes the `sandbox` key useful for including a third-party library that relies on `eval()` or `new Function()`: load the library in a sandboxed page, and use `postMessage()` to send it data from, and return results to, the rest of the extension.

## Syntax

`sandbox` is an object with these properties:

<dl>
  <dt><code>pages</code></dt>
  <dd>
    Required. An array of strings. A list of paths, relative to manifest.json, to the extension's sandboxed pages.
  </dd>
  <dt><code>content_security_policy</code></dt>
  <dd>
    Optional. A string. <strong>Manifest V2 only.</strong> The content security policy applied to the sandboxed pages. For the Manifest V3 equivalent, see the <code>sandbox</code> property of the <a href="/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/content_security_policy"><code>content_security_policy</code></a> key, described in <a href="#content_security_policy_for_sandboxed_pages">Content security policy for sandboxed pages</a>.
  </dd>
</dl>

> ```json

Reviewer: Two examples without context doing different things are not ideal. Could you move this to the explanatory section below, starting with the MV3 example, and then a note explaining what it does? I added a suggested edit.

// Manifest V2
"sandbox": {
  "pages": ["sandbox.html"],
  "content_security_policy": "sandbox allow-scripts; script-src 'self' 'unsafe-eval';"
}
```

```json
// Manifest V3
"sandbox": {
  "pages": ["sandbox.html"]
}
```

> ## Content security policy for sandboxed pages
> 
> In Manifest V3, the content security policy for sandboxed pages is set using the `sandbox` property of the [`content_security_policy`](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/content_security_policy) key. **This property is Manifest V3 only.** In Manifest V2, use the `content_security_policy` property of the `sandbox` key instead, as described in [Syntax](#syntax) above.
> 
> ```json
> "content_security_policy": {
>   "sandbox": "sandbox allow-scripts; script-src 'self' 'unsafe-eval';"
> }
> ```

Reviewer: I think that it might be more clear to put the equivalent manifest files after each other, instead of splitting them over separate sections. And also one explanation on what the CSP is doing:

````suggestion

For example, to customize the CSP to enable `eval()` in sandbox.html, add `'unsafe-eval'` as follows:

```json
// Manifest V3
"sandbox": {
  "pages": ["sandbox.html"]
},
"content_security_policy": {
  "sandbox": "sandbox allow-scripts; script-src 'self' 'unsafe-eval';"
}
```

```json
// Manifest V2
"sandbox": {
  "pages": ["sandbox.html"],
  "content_security_policy": "sandbox allow-scripts; script-src 'self' 'unsafe-eval';"
}
```
````

If a policy is not supplied, sandboxed pages get this default content security policy:

```plain
sandbox allow-scripts; script-src 'self';
```

This isolates a sandboxed page from the rest of the extension, but doesn't allow `eval()` or similar constructs. To permit these, include `'unsafe-eval'` in the `script-src` directive of a custom policy.

Any custom policy supplied for sandboxed pages must meet these requirements:

- It must include the {{CSP("sandbox")}} directive.
- The {{CSP("sandbox")}} directive must not include the `allow-same-origin` keyword. Allowing this would give the page access to the rest of the extension's origin, defeating the purpose of sandboxing it.

## Best practices

When using a sandboxed page, consider these best practices:

> - **Validate messages.** As a sandboxed page communicates with the rest of the extension using {{DOMxRef("Window.postMessage()")}}, it's exposed to the same risks as any other use of `postMessage()`. Check the origin and shape of every message before acting on it. See [security concerns](/en-US/docs/Web/API/Window/postMessage#security_concerns) in the `postMessage()` documentation.

Reviewer: ```suggestion
- **Validate messages.** If a sandboxed page communicates with the rest of the extension using {{DOMxRef("Window.postMessage()")}}, it's exposed to the same risks as any other use of `postMessage()`. Check the origin and shape of every message before acting on it. See [security concerns](/en-US/docs/Web/API/Window/postMessage#security_concerns) in the `postMessage()` documentation.
```

> - **Validate messages.** As a sandboxed page communicates with the rest of the extension using {{DOMxRef("Window.postMessage()")}}, it's exposed to the same risks as any other use of `postMessage()`. Check the origin and shape of every message before acting on it. See [security concerns](/en-US/docs/Web/API/Window/postMessage#security_concerns) in the `postMessage()` documentation.

Reviewer: Perhaps we should clarify what "origin" checking means. My explanation is scattered over comments in this PR.

Regular extension pages have the usual `"moz-extension://uuid"` origin (same as `location.origin`, which is the same between sandboxed pages and extension pages). Sandboxed pages expecting messages from itself should validate that `event.origin` is the extension origin, and messages targeting the extension page should use the same. That fits with the best practices documented in the MDN link.

Sandboxed pages have an opaque origin serializing as `"null"`, so messages received from it have `event.origin` set to "null", and messages targeting it can only use `"*"` (against the best practice from the linked postMessage page; but there is no alternative until this spec issue is resolved: https://github.com/whatwg/html/issues/3585 ).

## Example

This example loads a third-party templating library that uses `eval()`-like constructs into a sandboxed page, and uses it from a popup.

`manifest.json`:

```json
{
  "manifest_version": 3,
  "name": "Sandbox example",
  "version": "1.0",
  "action": {
    "default_popup": "popup.html"
  },
  "sandbox": {
    "pages": ["sandbox.html"]
  },
  "content_security_policy": {
    "sandbox": "sandbox allow-scripts; script-src 'self' 'unsafe-eval';"
  }
}
```

`sandbox.html`:

```html
<!doctype html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <script src="templating-library.js"></script>
    <script src="sandbox.js"></script>
  </head>
  <body></body>
</html>
```

`templating-library.js` stands in for a third-party library that compiles templates using `new Function()`, which is why it must run in a sandboxed page:

```js
const TemplatingLibrary = {
  compile: (template) => (context) =>

>     new Function(...Object.keys(context), `return \`${template}\`;`)(

Reviewer: This example is hard to understand. I also asked elsewhere to drop "templating" as the motivating use case across the patch. Can we make a contrived example? Maybe a calculator plus eval?

      ...Object.values(context),
    ),
};
```

`sandbox.js` listens for messages from the popup, renders a template using the sandboxed library, and posts the result back:

```js
window.addEventListener("message", (event) => {
  if (event.source !== window.parent) {
    // Ignore messages from any window other than the
    // extension page that embeds this sandboxed page.
    // The sandboxed page has an opaque origin, so
    // location.origin can't be used for this check.
    return;
  }
  const { template, context } = event.data;
  const render = TemplatingLibrary.compile(template);
  event.source.postMessage({ result: render(context) }, event.origin);
});
```

> `popup.html` embeds the sandboxed page in a hidden iframe. The `defer` attribute on `popup.js` ensures it only runs after the iframe has been parsed:

Reviewer: Drop the defer mention; it is not relevant to the feature. If you wanted to run the code later, just put the tag after the iframe element.

But I'd also change the example to not use a HTML iframe, but create the frame with createElement. Then you can create the iframe element and attach a message listener, before appending the frame to the document. That ensures that you won't miss a message event.

Something like this (untested). Well after writing the example I decided to not register the message handler early, because the code example looks more linear/understandable in this way. But the pattern is more easy to adapt to the scenario where one wants to handle messages before postMessage.

```javascript
async function runCalculation(expr) {
  const iframe = document.createElement("iframe");
  iframe.src = "sandbox.html";
  iframe.hidden = true;

  await new Promise(resolve => {
    iframe.addEventListener("load", resolve, { once: true });
    document.body.append(iframe);
  });

  let replyPromise = new Promise(resolve => {
    const handleMsg = event => {
      if (event.source !== iframe.contentWindow) {
        return;
      }
      window.removeEventListener("message", handleMsg);
      resolve(event.data);
    });
    window.addEventListener("message", handleMsg);
  });
  // Best practice is to use the specific origin instead of "*",
  // but a sandboxed frame has origin "null", so
  // we cannot send a more specific target origin.
  iframe.contentWindow.postMessage(expr, "*");
  const reply = replyPromise;

  iframe.remove();
  return reply;
}
```

}
```

```json
// Manifest V2
"sandbox": {
  "pages": ["sandbox.html"],
  "content_security_policy": "sandbox allow-scripts; script-src 'self' 'unsafe-eval';"
}
```

Any custom policy supplied for sandboxed pages must meet these requirements:

- It must include the {{CSP("sandbox")}} directive.
- The {{CSP("sandbox")}} directive must not include the `allow-same-origin` keyword. Allowing this would give the page access to the rest of the extension's origin, defeating the purpose of sandboxing it.

## Best practices

When using a sandboxed page, consider these best practices:

- **Validate messages.** If a sandboxed page communicates with the rest of the extension using {{DOMxRef("Window.postMessage()")}}, it's exposed to the same risks as any other use of `postMessage()`. Check the origin and shape of every message before acting on it. See [security concerns](/en-US/docs/Web/API/Window/postMessage#security_concerns) in the `postMessage()` documentation.

  A sandboxed page's `location.origin` reflects the extension's `moz-extension://` origin. Therefore, a sandboxed page can validate an incoming message by checking that `event.origin` matches `location.origin`.

>   Messages sent _from_ a sandboxed page report an opaque origin (`"null"`). Therefore, the extension page receiving the messages can't validate by origin; instead, compare `event.source` to the sandboxed frame's `contentWindow`.

Reviewer: ```suggestion
  Messages sent _from_ a sandboxed page report the serialization of opaque origin (`"null"`), which is indistinguishable from other opaque origins. Therefore, the extension page receiving the messages can't validate by origin; instead, compare `event.source` to the sandboxed frame's `contentWindow`.
```
